### ISSUE TITLE
Firefox shows "attempting to reconnect to server" on opening application; 

### ISSUE THREAD

**Issue by Stanislav-Beresnev** [dotnet/aspnetcore#28592]
### Describe the bug
I have the same issue as that [https://github.com/dotnet/aspnetcore/issues/16739](https://github.com/dotnet/aspnetcore/issues/16739) but in my case "attempting to reconnect to server" is displayed when I open the application or press F5 to refresh the page. Sometimes, I faced it when navigating to another page. See the attached picture.

![Firefox_bug](https://user-images.githubusercontent.com/12596197/101907014-f1bc1580-3bca-11eb-8e45-d85f367472d6.jpg)

### To Reproduce

1. Using ASP.NET core 3.0/3.1 or .NET 5.0
2. Create a new server side blazor project
3. Run the program in firefox

### Expected behavior
The "attempting to reconnect to server" error message is not showed.
### Further technical details
- ASP.NET core 3.0/3.1 or .NET 5.0
- Microsoft Visual Studio Professional 2019 Version 16.8.1, Firefox v. 83.0 (64bit)

**Comment by javiercn** [comment]
@Stanislav-Beresnev thanks for contacting us.

@captainsafia do you have any ideas here? We believe this is similar to something you worked on recently.

**Comment by captainsafia** [comment]
@Stanislav-Beresnev Can you try to repro the issue on the latest version of .NET 5 and ASP.NET 3.2 respectively? I'm unable to repro on the latest version with Firefox 81.0.

We shipped some fixes to the disconnect process in the latest patch releases.

**Comment by Stanislav-Beresnev** [comment]
@captainsafia I can still reproduce this, but I used the latest version of Firefox - 84.0. However, now I can only see this message after I refresh the page several times (5-10).

**Comment by ponchautf** [comment]
Hi,
I also have this issue with Firefox (v84.0.2) on Mac OS.
My app is a classic Asp.net core MVC application (.net core 3.1 using controller & cshtml views) with some server blazor components.

I can clearly see the message each time I navigate to another View. It's displayed less than a second but it's annoying.
Is there a way to avoid this?

this don't occurs on Google chrome (not yet tested on edge or safari)

**Comment by mikoskinen** [comment]
Started seeing this today. Didn't happen yesterday (Firefox 97.0). What changed was that I changed the project to target .NET 6 instead of Core 3.1. 

Still doesn't happen with Edge but happens every time when I click a link with Firefox. It seems that if it takes the server 1 second to process the request, I see the "attempting to reconnect to server" for that second.

**Comment by Bouke** [comment]
I'm posting here as @mkArtakMSFT marked my bug reports as duplicate of this one. I've found that there are two separate issues w.r.t. Firefox that cause this dialog to appear, both solved with the same workaround.

# Issue 1: Firefox temporarily shows reconnect dialog while navigating to new page (or reloading page)

On Firefox the reconnect dialog temporarily appears while navigating to a new slow page and it disappears while the browser is still waiting on a response from the server. For example on a page that takes 3 seconds to render on the server side:

1. Browser starts loading new page
2. Reconnect dialog appears immediately
3. After ~1 second the reconnect dialog disappears
4. After ~3 seconds the new page appears

This is very confusing for the user, as the reconnect dialog makes no sense in the context of the page navigation: once the reconnect dialog appears the browser is still waiting for the new page.

I _think_ the issue is with Firefox closing the websocket as soon as the browser starts the navigation. Blazor is still running on the current page and tries to reconnect. Then when the new page is loaded that connection is then dropped again. I believe the following log displays this:

# Issue 2: Reconnect modal displays immediately in Firefox upon navigation

The problem seems to be that Firefox executes both `this.modal.style.visibility="hidden"` and `this.modal.style.visibility="visible"` in the same reflow. As a result, the css transition is not triggered. When I change the `setTimeout` delay from 0 to 10 ms the transition is triggered and the delay of 500ms (+10) is respected.

# Workaround

A custom reconnect dialog with a longer delay works around both issues:

- The longer delay (e.g. 5 seconds) hides the reconnect that happens during page unload.
- The custom reconnect dialog will already be part of the DOM, so the CSS transition delay is applied correctly on unload.

**Comment by vindberg** [comment]
It seems like the reconnection issue with Blazor is growing! Can you please consider fixing this vital error in version 7.0! Im so sad to see it pushed to 8.

**Comment by Liebeck** [comment]
I ran into the same issue today. Seems to be fixed with

`        services.AddServerSideBlazor()
            .AddHubOptions(options =>
            {
                options.ClientTimeoutInterval = TimeSpan.FromSeconds(5);
            });`

**Comment by obiwanconobi** [comment]
> I ran into the same issue today. Seems to be fixed with
> 
> ```
>  services.AddServerSideBlazor()
>             .AddHubOptions(options =>
>             {
>                 options.ClientTimeoutInterval = TimeSpan.FromSeconds(5);
>             });
> ```

This didn't work for me

**Comment by ponchautf** [comment]
Sometime these error are due to 500 server error.
Debug your app (make sure that you stop on any errors’ not just your code)
You will probably find that an exception is raised somewhere

**Comment by zacuke** [comment]
Hello, I am experiencing this intermittent reconnect dialog when the server is not having any connection issues. It seems to happen mostly in Firefox, especially on logout. In general this makes Blazor a difficult framework to recommend to management as they don't have an appreciation of the finer details regarding Firefox having poor support for the onBeforeUnload event, or that a Websocket is being closed too early, or some kind of CSS race condition is happening, or any of that kind of stuff. I should have created a plain Razor pages app, and I might have to go back to that, but I'm pushing forward with Blazor as I do love the technology from a theoretical stand point.

The suggested workarounds above don't seem to work, or maybe I don't understand them correctly.  `options.ClientTimeoutInterval = TimeSpan.FromSeconds(5);` definitly isn't correct, as that reduces the ClientTimeoutInternal from 30 to 5 seconds and causes the reconnect panel to flash much more often.

Regarding `Firefox executes both this.modal.style.visibility="hidden" and this.modal.style.visibility="visible" in the same reflow.` - exactly what CSS or JavaScript should be changed? Are we to implement a custom `CircuitHandler` and add it via `.AddCircuitHandler<CustomCircuitHandler>();` 

In summary, could somebody provide a more comprehensive workaround guide? I think most of us don't mind the extra work to get some cool tech like Blazor working reliably, but the default reconnect splash message is a real black eye on an otherwise amazing framework.

Thanks to all for the continued discussion regarding this long standing annoyance.

**Comment by zacuke** [comment]
I had already upgraded the project sdk reference to .net 8 a few weeks ago, but today I realized I am still on old style Blazor Server. So I also upgraded to Blazor web app per these instructions: https://learn.microsoft.com/en-us/aspnet/core/migration/70-80?view=aspnetcore-8.0&tabs=visual-studio#update-a-blazor-server-app

This cleans up a lot of the behavior of things and it seems what I was trying to do is built-in now via `@rendermode InteractiveServer`

[Bouke](https://github.com/Bouke), thanks for that information I will look into having a custom reconnect modal to resolve the issue with Firefox still showing a Reconnect error on page navigation.

**Comment by javiercn** [comment]
The original issue is for a much older version of Blazor, and it is likely not applicable any longer.

In .NET 10.0 the reconnect dialog is part of the default template and can be customized/tweaked as needed.

There seem to be multiple reports about this, but all from different versions of Blazor and under different conditions, which makes it really hard for us to understand or try and reproduce this issue.

If you are still experiencing problems, we suggest the following:
* Take note of any errors in the console or the server logs (debug level).
* Clearly explain the situation in which this happens for you (are you navigating away, an exception happens on the server, and so on).
* Use the reconnect modal from .NET 10.0 and see if that solves the problem.
* If you still have problems after that and have come up with a concrete CSS/html-based solution that is applicable here, feel free to propose a change.
